morello/src/search.rs

```diff
@@ -702,69 +702,59 @@ impl<Tgt: Target> SpecTask<Tgt> {
 impl ImplReducer {
     fn new(top_k: usize, preferences: SmallVec<[ActionIdx; 1]>) -> Self {
         ImplReducer {
-            results: smallvec![],
+            results: BTreeSet::new(),
             top_k,
             preferences,
         }
     }
 
-    fn insert(&mut self, new_action_idx: ActionIdx, cost: Cost) {
-        match self.results.binary_search_by_key(&&cost, |imp| &imp.1) {
-            Ok(idx) => {
-                debug_assert!(idx < self.top_k);
-                // Replace something if it improves preference count, and do
-                //   nothing if not.
-                let mut to_set = None;
-                for i in self.iter_surrounding_matching_cost_indices(idx, &cost) {
-                    // TODO: Instead of filtering here, just push down the length.
-                    if i >= self.preferences.len() {
-                        continue;
-                    }
-
-                    if new_action_idx == self.preferences[i] {
-                        to_set = Some(i);
-                        break;
-                    }
-                }
-                if let Some(i) = to_set {
-                    self.results[i].0 = new_action_idx;
-                }
+    fn insert(&mut self, new_action_idx: ActionIdx, new_cost: Cost) {
+        if self.results.len() < self.top_k {
+            // We have not yet filled the top_k, so just insert.
+            self.results.insert((new_cost, new_action_idx));
+        } else if self
+            .results
+            .iter()
+            .find(|&(cost, _)| *cost == new_cost)
+            .is_some()
+        {
+            debug_assert!(self.results.len() == self.top_k);
 
-                debug_assert!(self.results.len() <= self.top_k);
-            }
-            Err(idx) if idx < self.top_k => {
-                if self.results.len() == self.top_k {
-                    self.results.pop();
-                }
-                self.results.insert(idx, (new_action_idx, cost));
+            // We have filled the top_k and found the same cost in results, so
+            //   replace something if it improves preference count, and do
+            //   nothing if not.
+            if let Some(action) = self
+                .results
+                .iter()
+                .enumerate()
+                // Since we know that results is sorted by Cost, this filter
+                //   only takes contiguous elements with the same cost.
+                .filter(|&(i, (cost, _))| i < self.preferences.len() && *cost == new_cost)
+                .find(|&(i, _)| new_action_idx == self.preferences[i])
```

Comment:
@samkaufman  I think we can do something similar to the following, but the worst case could still be O(k). Or maybe I am misunderstanding what you want to do?

```rust
.range((new_cost.clone(), 0)..=(new_cost.clone(), ActionIdx::MAX))
```

Reply:
That makes sense.  Thank you!

Reply:
@samkaufman  Because `range` is only provided for `BTreeSet` (not `Iterator`) and we cannot take the original index, I think we cannot later compare `new_action_idx` and `self.preferences[i]` (i.e., the following is incorrect). Would you happen to have any idea how to achieve the correct implementation with `.range`?

```rs
                .results
                .range((new_cost.clone(), 0)..=(new_cost.clone(), ActionIdx::MAX))
                .enumerate()
                .find(|&(i, _)| new_action_idx == self.preferences[i])
```